packages/js/api-client/src/auth.ts

```diff
@@ -0,0 +1,191 @@
+import type { Middleware } from "openapi-fetch"
+
+import type { components } from "./generated/openverse"
+
+import type { OpenverseClient, ClientCredentials } from "./types"
+
+type OAuth2Token = components["schemas"]["OAuth2Token"]
+
+const currTimestamp = (): number => Math.floor(Date.now() / 1e3)
+export const EXPIRY_THRESHOLD = 5 // seconds
+
+export class OpenverseAuthMiddleware implements Middleware {
+  /**
+   * An Openverse REST API client.
+   *
+   * Used to perform the authentication workflow.
+   */
+  client: OpenverseClient
+
+  /**
+   * Credentials for the Openverse API.
+   */
+  credentials: ClientCredentials
+
+  /**
+   * The most recently retrieved API token.
+   */
+  apiToken: OAuth2Token | null = null
+
+  /**
+   * Represents the requesting state of the auth middleware.
+   *
+   * When `null`, no request is underway. When it is a Promise,
+   * the Promise will resolve with `true` once the API token is available,
+   * or `false` if an error occurs.
+   *
+   * If the token request fails, it will throw.
+   *
+   * This is immediately set back to `null` once the request cycle is complete.
+   */
+  requesting: null | Promise<boolean> = null
+
+  /**
+   * The most recent API token request failure, if any.
+   */
+  failure = null as unknown
+
+  /**
+   * The UNIX timestamp at which the current token will expire.
+   */
+  tokenExpiry: number | null = null
+
+  constructor(client: OpenverseClient, credentials: ClientCredentials) {
+    this.client = client
+    this.credentials = credentials
+  }
+
+  onRequest: Middleware["onRequest"] = async ({ schemaPath, request }) => {
+    if (schemaPath == "/v1/auth_tokens/token/") {
+      // Do not send auth headers for token generation requests
+      return request
+    }
+
+    const apiToken = await this.getApiToken()
+    request.headers.set("Authorization", `Bearer ${apiToken}`)
+    return request
+  }
+
+  /**
+   * Get the timestamp as the number of seconds from the UNIX epoch.
+   * @returns the UNIX timestamp with a resolution of one second
+   */
+  get currTimestamp() {
+    return Math.floor(Date.now() / 1e3)
+  }
+
+  /**
+   * Determine if the current state of the client requires waiting for
+   * a token refresh to complete. This is distinct from whether
+   * the client should trigger a token refresh, because if an API token
+   * already exists, the refresh is eagerly requested before the current
+   * token expires. That means that if the request triggering the token refresh
+   * happens within the expiry threshold (see `expiryThreshold`), it
+   * can proceed with the existing token, as it is still valid.
+   *
+   * However, if the token is expired (or no token yet exists), the request
+   * must be held until the client has an API token.
+   *
+   * This is only relevant to clients configured with credentials. Non-credentialed
+   * clients bypass this process altogether.
+   *
+   * @returns Whether the client should wait.
+   */
```

Comment:
Understanding this flow might be helped by reading the tests that exercise it, so check there first if you aren't following the approach.

As mentioned in the PR description, the same methodology is proven in the frontend's `api-token` plugin, except there it blocks the full SSR request until the token returns, rather than just individual API requests. The principle and effect is the same, and I managed to sort out an implementation that works without needing to rely on the third-party async-mutex library like we use in the frontend.